pl_examples/loops_customisation/k_fold.py

```diff
@@ -0,0 +1,216 @@
+
+    def setup(self, stage: Optional[str] = None) -> None:
+        if stage == "fit" or stage is None:
+            self._train_dataset = Subset(self.random_full, indices=range(64))
+            self.dims = self._train_dataset[0].shape
+
+        if stage in ("fit", "validate") or stage is None:
+            self._val_dataset = Subset(self.random_full, indices=range(64, 64 * 2))
+
+        if stage == "test" or stage is None:
+            self._test_dataset = Subset(self.random_full, indices=range(64 * 2, 64 * 3))
+            self.dims = getattr(self, "dims", self._test_dataset[0].shape)
+
+        if stage == "predict" or stage is None:
+            self._predict_dataset = Subset(self.random_full, indices=range(64 * 3, 64 * 4))
+            self.dims = getattr(self, "dims", self._predict_dataset[0].shape)
+
+
+class KFoldLoop(ExternalLoop):
+    def __init__(
+        self,
+        num_folds: int,
+        best_model_paths: List[str] = [],
+        restarting: bool = False,
+    ):
+        super().__init__()
+        self.num_folds = num_folds
+        self.best_model_paths = best_model_paths
+        self.restarting = restarting
+
+    @staticmethod
+    def loop_base_callback() -> Type[Callback]:
+        class BaseKFoldCallback(Callback):
+            @rank_zero_only
+            def on_fold_start(self, trainer, pl_module, counter):
+                """Override with your own logic"""
+
+        return BaseKFoldCallback
+
+    @property
+    def done(self) -> bool:
+        return self.current_fold >= self.num_folds
+
+    def reset(self) -> None:
+        if not self.restarting:
+            self.current_fold = 0
+
+    def on_run_start(self, *args: Any, **kwargs: Any) -> None:
+        # temporary hack
+        self.trainer.datamodule.setup("fit")
+
+    def on_advance_start(self) -> None:
+        # more reproducible as re-creating a different trainer.
+        self.create_trainer(max_epochs=np.random.randint(10))
+        # reload dataset for the current fold
+        dm = self.trainer.datamodule
+        dm.processed_train_dataset = self.process_dataset("train", dm.train_dataset)
+        dm.processed_val_dataset = self.process_dataset("val", dm.val_dataset)
+        # call user hook
+        self.trainer.call_hook("on_fold_start", self.current_fold)
+        # reset model parameters
+        self.trainer.lightning_module.reset_parameters()
+
+    def advance(self) -> Any:
+        # dataloaders will be automatically reloaded
+        return self.trainer.fit(self.trainer.lightning_module, datamodule=self.trainer.datamodule)
+
+    def on_advance_end(self) -> None:
+        self.current_fold += 1
+        # stored best weight path for this fold
+        self.best_model_paths.append(self.trainer.checkpoint_callback.best_model_path)
+
+    # utilities for creating a hold
+    def process_dataset(self, stage: str, dataset: Dataset) -> Subset:
+        kfold = KFold(self.num_folds, random_state=42, shuffle=True)
```

Comment:
is a dependency for sklearn worth it for just this?
Should we maybe have a more general abstract function `create_splits` the user has to implement? Since there are so many different ways to create data splits.